src/live_restore/live_restore_fs.c

```diff
@@ -246,18 +248,94 @@ __live_restore_fs_find_layer(WT_FILE_SYSTEM *fs, WT_SESSION_IMPL *session, const
     return (0);
 }
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
 /*
- * __live_restore_fs_notsup --
- *     Return an error message indicating the given functionality is not supported.
+ * __live_restore_fs_directory_list_worker --
+ *     The list is a combination of files from the destination and source directories. For
+ *     destination files, exclude any files matching the marker paths. For source files, exclude
+ *     files that are either marked as tombstones or already present in the destination directory.
  */
 static int
-__live_restore_fs_notsup(WT_SESSION *wt_session)
+__live_restore_fs_directory_list_worker(WT_FILE_SYSTEM *fs, WT_SESSION *wt_session,
+  const char *directory, const char *prefix, char ***dirlistp, uint32_t *countp, bool single)
 {
-    WT_RET_MSG((WT_SESSION_IMPL *)wt_session, ENOTSUP, "Unsupported fs operation");
+    WT_DECL_RET;
+    WT_LIVE_RESTORE_FS *lr_fs;
+    WT_SESSION_IMPL *session;
+    size_t dirallocsz;
+    uint32_t count_dest, count_src;
+    char **dirlist_dest, **dirlist_src, **entries, **namep, *path_dest, *path_src;
+    bool dest_exist, have_tombstone;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    lr_fs = (WT_LIVE_RESTORE_FS *)fs;
+    count_dest = count_src = dirallocsz = 0;
+    *dirlistp = dirlist_dest = dirlist_src = entries = NULL;
+    path_dest = path_src = NULL;
+    dest_exist = have_tombstone = false;
+
+    __wt_verbose_debug1(session, WT_VERB_FILEOPS, "DIRECTORY LIST %s (single ? %s) : ", directory,
+      single ? "YES" : "NO");
+
+    /* Get files from destination. */
+    WT_ERR(__live_restore_fs_backing_filename(
+      &lr_fs->destination, session, lr_fs->destination.home, directory, &path_dest));
+    WT_ERR(lr_fs->os_file_system->fs_directory_list(
+      lr_fs->os_file_system, wt_session, path_dest, prefix, &dirlist_dest, countp));
+
+    for (namep = dirlist_dest; *namep != NULL; namep++)
+        if (namep != NULL && *namep != NULL &&
+          !(strlen(*namep) >= strlen(".deleted") &&
```

Comment:
We know that `*namep != NULL` per the loop check so we don't need to check that in the if statement here. I also think the check for a ".deleted" should be simplified and possibly turned into a function like __name_is_tombstone(char *filename)

Reply:
+1
This boolean is also pretty difficult to parse can we either store the result in a clearly named bool
```
is_marker_file = (namep != NULL && *namep != NULL &&
          !(strlen(*namep) >= strlen(".deleted") && strcmp(*namep + strlen(*namep) - strlen(".deleted"), ".deleted") == 0);
if (!is_marker_file) {
```
Or add a clarifying comment?